hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableKVStateBase.java

```diff
@@ -0,0 +1,265 @@
+     * be called by the code that created the {@link WritableKVStateBase} instance or owns it. Don't
+     * cast and commit unless you own the instance!
+     */
+    public final void commit() {
+        for (final var entry : modifications.entrySet()) {
+            final var key = entry.getKey();
+            final var value = entry.getValue();
+            if (value == null) {
+                removeFromDataSource(key);
+            } else {
+                putIntoDataSource(key, value);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Clears the set of modified keys and removed keys. Equivalent semantically to a "rollback"
+     * operation.
+     */
+    @Override
+    public final void reset() {
+        super.reset();
+        modifications.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V get(@NonNull K key) {
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            final var value = modifications.get(key);
+            super.markRead(key, value);
+            return value;
+        } else {
+            return super.get(key);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V getForModify(@NonNull final K key) {
+        Objects.requireNonNull(key);
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            return modifications.get(key);
+        }
+
+        // If the modifications map does not contain an answer, but the read cache of the
+        // super class does, then it means we've looked this up before but never modified it.
+        // So we can just delegate to the super class.
+        if (hasBeenRead(key)) {
+            return super.get(key);
+        }
+
+        // We have not queried this key before, so let's look it up and store that we have
+        // read this key. And then return the value.
+        final var val = getForModifyFromDataSource(key);
+        markRead(key, val);
+        return val;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public final void put(@NonNull final K key, @NonNull final V value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        modifications.put(key, value);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public final void remove(@NonNull final K key) {
+        Objects.requireNonNull(key);
+        modifications.put(key, null);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>This is a bit of a pain, because we have to take into account modifications! If a key has
+     * been removed, then it must be omitted during iteration. If a key has been added, then it must
+     * be added into iteration.
+     *
```

Comment:
We should probably add a comment that this `Iterator` is not fail-fast.

Reply:
Done